password_enhancements.module

```diff
@@ -0,0 +1,471 @@
+  _password_enhancements_add_password_requirement_validation($form, $form_state);
+  $form['password_enhancements_password_change_required']['#access'] = Drupal::currentUser()->hasPermission('administer user password settings');
+}
+
+/**
+ * Implements hook_form_FORM_ID_alter().
+ */
+function password_enhancements_form_password_enhancements_change_form_alter(&$form, FormStateInterface $form_state, $form_id): void {
+  _password_enhancements_add_password_requirement_validation($form, $form_state);
+}
+
+/**
+ * Implements hook_form_FORM_ID_alter().
+ */
+function password_enhancements_form_user_register_form_alter(array &$form, FormStateInterface $form_state, string $form_id): void {
+  _password_enhancements_add_password_requirement_validation($form, $form_state);
+}
+
+/**
+ * Adds password requirement validation.
+ *
+ * @param array $form
+ *   The form for which the password requirements validation should be added.
+ * @param \Drupal\Core\Form\FormStateInterface $form_state
+ *   The current form state.
+ */
+function _password_enhancements_add_password_requirement_validation(array &$form, FormStateInterface $form_state): void {
+  if (empty($form['account']['pass'])) {
+    /** @var \Drupal\Core\Config\Config $config */
+    $config = Drupal::config('password_enhancements.settings');
+    if (empty($config->get('require_password'))) {
+      // No password field available, nothing to do here.
+      return;
+    }
+
+    $form['account']['pass'] = [
+      '#type' => 'password_confirm',
+      '#size' => 25,
+      '#description' => t('Provide a password for the new account in both fields.'),
+      '#required' => TRUE,
+    ];
+    $form['actions']['submit']['#submit'][] = 'password_enhancements_form_user_register_form_alter_password_required_submit';
+  }
+
+  // Add weight for account fields.
+  $weight = 0;
+  foreach (Element::children($form['account']) as $key) {
+    if (!array_key_exists('#weight', $form['account'][$key])) {
+      $form['account'][$key]['#weight'] = $weight++;
+    }
+  }
+
+  // Get roles.
+  $roles = $form_state->getValue('roles') ?? [];
+  $input = $form_state->getUserInput();
+  // If there was no input then grab the default roles.
+  if (empty($roles) && empty($input['roles']) && !empty($form['account']['roles']['#default_value'])) {
+    $roles = $form['account']['roles']['#default_value'];
+  }
+  $roles = array_merge([AccountInterface::AUTHENTICATED_ROLE], $roles, (!empty($input['roles']) ? array_values($input['roles']) : []));
+
+  try {
+    /** @var \Drupal\password_enhancements\Entity\Storage\ConstraintEntityStorageInterface $constraint_storage */
+    $constraint_storage = Drupal::entityTypeManager()->getStorage('password_enhancements_constraint');
+    $constraints = $constraint_storage->loadByRoles($roles);
+  }
+  catch (Exception $e) {
+    Drupal::service('logger.password_enhancements')->exception($e);
+    $constraints[] = new Constraint([
```

Comment:
The logic in this function seems a little bit chaotic to me. You are only calling this function in form alter hooks.
Why don't you just set all form elements `#access => FALSE` on the `$form` and display an error message if something unexpected occurs? Because I guess when `An error occurred, it is currently not possible to validate the password` occurs the user should not be able to submit the form.
Constructing these "magical" `error` type constraint _entities_ seems to be just a workaround to me because they never get saved to the database and they should not be saved if I am correct :)

Reply:
This is an "internal" plugin, so it does not appear on the portal anywhere except if there is an error, but it cannot be saved to the database.

I wanted to use the same validation process in case of an error as for the rest of the constraints, so I didn't had to do too much changes around it.

Also I think it is still better to display the error message in place rather than in the top of the page someone, and also this way it is validated too when the form is being submitted, so if the error still persist the user won't be able to change the password.